### PR TITLE
Remake dry-run=server optional

### LINKED ISSUE #536: `--color` flag doesn't work anymore
In v3.9.2. `--color` doesn't force a coloured output anymore.

- In 2.8.1 a command such as `helm diff upgrade NAME CHART --color | cat` would show a coloured result.
- In 2.9.2 the output is monochrome.

If the output is direct to TTY, 2.9.2 produces a coloured output as well. It's the `--color` flag to force it to seem broken.

### PR DESCRIPTION
**Pull request by mumoshu**
This is a follow-up on #499 which redoes it without relying on cobra's DisableFlagParsing. DisableFlagParsing turned out to break far many things than I imagined!

Fixes #536

Ref #537 #538

### CODE REVIEW

File: cmd/upgrade.go [hunk 1]
```diff
@@ -293,6 +231,9 @@ func newChartCommand() *cobra.Command {
 	f.BoolVar(&diff.devel, "devel", false, "use development versions, too. Equivalent to version '>0.0.0-0'. If --version is set, this is ignored.")
 	f.BoolVar(&diff.disableValidation, "disable-validation", false, "disables rendered templates validation against the Kubernetes cluster you are currently pointing to. This is the same validation performed on an install")
 	f.BoolVar(&diff.disableOpenAPIValidation, "disable-openapi-validation", false, "disables rendered templates validation against the Kubernetes OpenAPI Schema")
+	f.StringVar(&diff.dryRunMode, "dry-run", "", "--dry-run, --dry-run=client, or --dry-run=true disables cluster access and show diff as if it was install. Implies --install, --reset-values, and --disable-validation."+
+		" --dry-run=server enables the cluster access with helm-get and the lookup template function.")
+	f.Lookup("dry-run").NoOptDefVal = "true"
```
Comment (mumoshu) [inline cmd/upgrade.go:236]: So this was the biggest finding and improvement for me.
This is how helm kept the backward-compatibility on the flag when they introduced `dry-run=server` https://github.com/tapaskapadia/helm/blob/b7a2d47eca70e18b821b415efa5f47971c8a5302/cmd/helm/install.go#L166-L167

File: cmd/upgrade.go [hunk 2]
```diff
@@ -153,71 +150,12 @@ func newChartCommand() *cobra.Command {
 			"# Read the flag usage below for more information on --context.",
 			"HELM_DIFF_OUTPUT_CONTEXT=5 helm diff upgrade my-release datadog/datadog",
 		}, "\n"),
+		Args: func(cmd *cobra.Command, args []string) error {
+			return checkArgsLength(len(args), "release name", "chart path")
+		},
```
Comment (mumoshu) [inline cmd/upgrade.go:155]: We just moved this checkArgsLength from RunE to Args, as we used to do so before #499

File: cmd/upgrade.go [hunk 3]
```diff
@@ -89,7 +87,7 @@ func (d *diffCmd) isAllowUnreleased() bool {
 //
 // See also https://github.com/helm/helm/pull/9426#discussion_r1181397259
 func (d *diffCmd) clusterAccessAllowed() bool {
-	clientOnly := (d.dryRunModeSpecified && d.dryRunMode == "") || d.dryRunMode == "true" || d.dryRunMode == "client"
+	clientOnly := d.dryRunMode == "true" || d.dryRunMode == "client"
```
Comment (mumoshu) [inline cmd/upgrade.go:90]: We could do better, but the intention is; we ensure that `dryRunMode` ever takes any of `true`, `false`, `client` and `server` by doing `if diff.dryRunMode != "true" && diff.dryRunMode != "false" && diff.dryRunMode != "client" && diff.dryRunMode != "server" {` check in `cmd/upgrade.go`.

It's enough to check for `true` and `client` to say it's `clientOnly`.
If the user didn't told us to be `clientOnly`, it implies `clusterAccessAllowed`.

File: cmd/upgrade.go [hunk 4]
```diff
@@ -293,6 +242,9 @@ func newChartCommand() *cobra.Command {
 	f.BoolVar(&diff.devel, "devel", false, "use development versions, too. Equivalent to version '>0.0.0-0'. If --version is set, this is ignored.")
 	f.BoolVar(&diff.disableValidation, "disable-validation", false, "disables rendered templates validation against the Kubernetes cluster you are currently pointing to. This is the same validation performed on an install")
 	f.BoolVar(&diff.disableOpenAPIValidation, "disable-openapi-validation", false, "disables rendered templates validation against the Kubernetes OpenAPI Schema")
+	f.StringVar(&diff.dryRunMode, "dry-run", "", "--dry-run, --dry-run=client, or --dry-run=true disables cluster access and show diff as if it was install. Implies --install, --reset-values, and --disable-validation."+
+		" --dry-run=server enables the cluster access with helm-get and the lookup template function.")
+	f.Lookup("dry-run").NoOptDefVal = dryRunNoOptDefVal
```
Comment (mumoshu) [inline cmd/upgrade.go:247]: This was the most significant finding and improvement for me.
This is how helm kept the backward-compatibility on the flag when they introduced dry-run=server https://github.com/tapaskapadia/helm/blob/b7a2d47eca70e18b821b415efa5f47971c8a5302/cmd/helm/install.go#L166-L167

File: cmd/upgrade.go [hunk 5]
```diff
@@ -153,71 +159,14 @@ func newChartCommand() *cobra.Command {
...
-				// cmd.Args = func(cmd *cobra.Command, args []string) error {
-				//     return checkArgsLength(len(args), "release name", "chart path")
-				// }
-				//
-				// Because it seems to take precedence over the PersistentPreRunE
-				if err := checkArgsLength(len(args), "release name", "chart path"); err != nil {
-					return err
-				}
+			if diff.dryRunMode == "" {
+				diff.dryRunMode = "none"
```
Comment (mumoshu) [inline cmd/upgrade.go:167]: We can indeed use either `""` or `"none"` to denote that the dry run flag is unspecified.
However I intentionally chose `"none"` to make code consistent with helm. They use `"none"` to indicate unspecified dry-run too..

File: cmd/upgrade_test.go [hunk 6]
```diff
@@ -9,39 +9,51 @@ func TestIsRemoteAccessAllowed(t *testing.T) {
...
 			},
 			expected: false,
 		},
+		{
+			name: "legacy explicit dry-run=false flag",
+			cmd: diffCmd{
+				dryRunMode: "false",
+			},
+			expected: true,
+		},
```
Comment (mumoshu) [inline cmd/upgrade_test.go:31]: I missed adding the case for `false` in #499 so doing it here.

File: cmd/upgrade_test.go [hunk 7]
```diff
@@ -9,39 +9,51 @@ func TestIsRemoteAccessAllowed(t *testing.T) {
 		expected bool
 	}{
 		{
-			name:     "no flags",
-			cmd:      diffCmd{},
+			name: "no flags",
+			cmd: diffCmd{
+				dryRunMode: "none",
```
Comment (mumoshu) [inline cmd/upgrade_test.go:14]: As per https://github.com/databus23/helm-diff/pull/547/files#r1468284710

**Comment by yxxhero** [comment]
@mumoshu test locally. all is ok.

**Review by yxxhero — APPROVED** [review]

**Pull request merged**